### PR TITLE
Issue #3280811 by vnech: Book page in Flexible Group

### PR DESCRIPTION
**Pull request by nechai**
## Problem
There is no ability to add "Book page" nodes in "Flexible Group" groups.

## Solution
Add "Book page" node type as a group content to "Flexible Group" group type.

## Issue tracker
- https://www.drupal.org/project/social/issues/3280811
- https://getopensocial.atlassian.net/browse/YANG-7678

## Theme issue tracker
- [SocialBase](https://git.drupalcode.org/project/socialbase/-/merge_requests/127)
- [SociaBlue](https://git.drupalcode.org/project/socialblue/-/merge_requests/88)

## How to test
- [x] Login as an admin
- [x] Enable module `social_flexible_group_book`
- [x] Create a flexible group
- [x] In "Personalisation" section enable option "Enable book pages"
- [x] Visit page _/group/{gid}/books
- [x] Click on button "create book page" (make sure during creating you chose "create a new book" in "Book outline" section - otherwise the book page will not be visible on books overview page)
#### How to test stream notifications:
 - [x] Disable `social_advanced_queue` module (to make notifications creation faster)
 - [ ] When creating a book in FG in "Book outline" section chose "Create a new book" as notifications created only for new books without parents
 - [ ] Run cron after adding book in a FG

### Before merge
- [ ] Code/peer review is completed
- [ ] All commit messages are [clear and clean](https://open-social.slite.com/app/docs/DnmermZDIx_0OQ). If applicable a rebase was performed
- [ ] All automated tests are green
- [ ] Functional/manual tests of the acceptance criteria are approved
- [ ] All acceptance criteria were met
- [ ] If applicable (I.E. new hook_updates) the update path from previous versions (major and minor versions) are tested
- [ ] This pull request has all required labels (team/type/priority)
- [ ] This pull request has a milestone
- [ ] This pull request has an assignee (if applicable)
- [ ] Any front end changes are tested on all major browsers
- [ ] New UI elements, or changes on UI elements are approved by the design team
- [ ] New features, or feature changes are approved by the product owner

### After merge
- [ ] Code is tested on all branches that it has been cherry-picked
- [ ] Update hook number might need adjustment, make sure they have the correct order
- [ ] The Drupal.org ticket(s) are updated according to this pull request status

## Screenshots
Books page overview:
Screenshot 2022-05-23 at 18 37 26

Option for enabling/disabling books per group:
Screenshot 2022-05-31 at 17 27 40

## Release notes
*Add possibility to have "Book page" node type in "Flexible Group"*
- Enable module `social_flexible_group_book` if you want to have book pages in flexible groups;
- By default only group managers and platform managers are able to create book pages in flexible group;

## Change Record
N/A

## Translations
N/A

### CODE REVIEW

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+    if ($block->getPluginId() === 'group_add_book_block' && !$is_books_enabled) {
+      return AccessResult::forbidden();
+    }
+
+    if (
+      !$is_books_enabled &&
+      $block->getPluginId() === 'group_statistic_block' &&
+      // We need to hide the block on "view.group_books.page_group_books" route
+      // but due to this moment user gets "access denie" and previous route is
```
Comment (rollki) [inline modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module:152]: denied

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/config/install/views.view.group_books.yml [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+        gc__node:
+          id: gc__node
+          table: group_content_field_data
+          field: gc__node
+          relationship: none
+          group_type: group
+          admin_label: 'Group content Content'
+          required: false
+          group_content_plugins:
+            'group_node:book': 'group_node:book'
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/config/install/views.view.group_books.yml:359]: Fixed

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module (hunk 1 above)
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module]: Fixed

**Comment by nechai** [comment]
Hi @rollki ! Fixed, pls check.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Review by rollki — APPROVED** [review]
LGTM, let's wait for tests to be green.

**Comment by rollki** [comment]
Hey @Ressinel 
Could you please take a look at this PR?

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by rollki** [comment]
After adding notifications still LGTM :)

**Review by Ressinel — COMMENTED** [review]
Hi, @nechai I have some comments, see below, please.

1. There is no group name to which a book belongs and there is redundant left padding from the **Follow content** button.

d82c2849e95cc12795c48d910536dfda

2. The **group manager** (he is only a verified user outside of the group) doesn't see the **Create book** button on the books overview page.

c130db1df3840041ed2cddce6b956db6
d1be1f949a0b1d9be53a3ce2cf655ce8


3. The **group manager** has access to create a Book page in Group (`/group/{gid}/content/create/group_node%3Abook`) but doesn't see the **Book outline** section.

b521032e2dd05a297bb2a5c87db09b79

4. Book page creation notifications are not sent to group members. Can you add HTT for notifications?

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by nechai** [comment]
Hi @Ressinel ! Regarding your [notices](https://github.com/goalgorilla/open_social/pull/2931#pullrequestreview-999715121):
1. For now we want to have some MVP version and this will be fixed later
2. Fixed in the [commit](https://github.com/goalgorilla/open_social/pull/2931/commits/45b470636f719961641e48d6b784f77af71f1de0)
3. Fixed in the [commit](https://github.com/goalgorilla/open_social/pull/2931/commits/45b470636f719961641e48d6b784f77af71f1de0)
4. I updated HTT section, please check

**Comment by Ressinel** [comment]
Hi @nechai 

> 1. For now we want to have some MVP version and this will be fixed later

Then let's make this module experimental.

**Comment by nechai** [comment]
Hi @Ressinel ! Marked the module as experimental

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Review by Ressinel — CHANGES_REQUESTED** [review]

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.info.yml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+name: 'Social Flexible Group - Book pages'
+description: 'Provides an ability to have Book Page node type in flexible groups.'
+type: module
+core_version_requirement: ^9
+package: Social (Experimental)
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.info.yml:5]: Added

File: modules/social_features/social_book/social_book.install [hunk 4]
```diff
@@ -275,3 +275,17 @@ function social_book_update_10301() {
     user_role_grant_permissions($role_id, ['translate book node']);
   }
 }
+
+/**
+ * Hide book image field labels in view displays.
+ */
+function social_book_update_11400(): string {
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_book/social_book.install:282]: Fixed

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.install [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ */
+function social_flexible_group_book_install(): void {
+  // Make this module hooks fired after "book" module.
+  $extension_config = \Drupal::configFactory()->getEditable('core.extension');
+  $weight = $extension_config->get('module.book');
+  // Set the weight of this module.
+  module_set_weight('social_flexible_group_book', $weight + 1);
+
+  /** @var \Drupal\social_flexible_group_book\SocialFlexibleGroupBookInstallHelper $install_helper */
+  $install_helper = \Drupal::classResolver(SocialFlexibleGroupBookInstallHelper::class);
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.install:21]: Moved to `.install `file

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+
+/**
+ * Implements hook_entity_form_display_alter().
+ */
+function social_flexible_group_book_entity_form_display_alter(EntityFormDisplayInterface $form_display, array $context): void {
+  if (
+    $context['entity_type'] === 'node'
+    && $context['bundle'] === 'book'
+    && $form_display->getMode() === 'default'
+  ) {
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module:24]: Fixed

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    // Add "groups" field to the form.
+    /* @see \Drupal\social_group\SocialGroupSelectorWidgetConfigOverride::loadOverrides */
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module:26]: Fixed

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        'autocomplete' => FALSE,
+      ],
+    ]);
+  }
+
+  if (
+    $context['entity_type'] === 'group'
+    && $context['bundle'] === 'flexible_group'
+    && $form_display->getMode() === 'default'
+  ) {
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module:45]: Fixed

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  switch ($form_id) {
+    case 'group_flexible_group_add_form':
+    case 'group_flexible_group_edit_form':
+      // Move 'enable_books' field to the "Personalisation" field group.
+      if (isset($form['#fieldgroups']['group_personalisation'])) {
+        $form['#fieldgroups']['group_personalisation']->children[] = 'enable_books';
+        $form['#group_children']['enable_books'] = 'group_personalisation';
+      }
+      break;
+
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module:71]: https://www.drupal.org/docs/develop/standards/coding-standards#controlstruct

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+ */
+function social_flexible_group_book_block_access(Block $block, string $operation, AccountInterface $account): AccessResultInterface {
+  if ($operation !== 'view') {
+    return AccessResult::neutral();
+  }
+
+  if (
+    in_array($block->getPluginId(), ['group_hero_block', 'group_statistic_block'])
+    || $block->id() === 'socialblue_pagetitleblock_content'
+  ) {
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module:145]: Fixed

File: modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    // @todo This should be added to the block configuration from here.
```
[the remark this answers is not in the document]
Comment (nechai) [reply modules/social_features/social_group/modules/social_group_flexible_group/modules/social_flexible_group_book/social_flexible_group_book.module:146]: Removed redundant `@todo`

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr2931-othcwflffm8fyfjcokialzqloqckc4ri.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/6286b661a94aea8366d969e9

**Review by Ressinel — APPROVED** [review]
LGTM

**Comment by Ressinel** [comment]
🍒-picked to `11.4.x`

**Pull request merged**